src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -20,23 +20,53 @@ private BoundExpression BindIsPatternExpression(IsPatternExpressionSyntax node,
             TypeSymbol expressionType = expression.Type;
             if ((object)expressionType == null || expressionType.SpecialType == SpecialType.System_Void)
             {
-                expressionType = CreateErrorType();
                 if (!hasErrors)
                 {
                     // value expected
                     diagnostics.Add(ErrorCode.ERR_BadPatternExpression, node.Expression.Location, expression.Display);
                     hasErrors = true;
                 }
+
+                expression = BadExpression(expression.Syntax, expression);
+            }
+
+            BoundPattern pattern = BindPattern(node.Pattern, expression.Type, hasErrors, diagnostics);
+            hasErrors |= pattern.HasErrors;
+            return MakeIsPatternExpression(
+                node, expression, pattern, GetSpecialType(SpecialType.System_Boolean, diagnostics, node),
+                hasErrors, diagnostics);
+        }
+
+        private BoundExpression MakeIsPatternExpression(SyntaxNode node, BoundExpression expression, BoundPattern pattern, TypeSymbol boolType, bool hasErrors, DiagnosticBag diagnostics)
+        {
+            // Note that these labels are for the convenience of the compilation of patterns, and are not actually emitted into the lowered code.
+            LabelSymbol whenTrueLabel = new GeneratedLabelSymbol("isPatternSuccess");
+            LabelSymbol whenFalseLabel = new GeneratedLabelSymbol("isPatternFailure");
+            BoundDecisionDag decisionDag = DecisionDagBuilder.CreateDecisionDagForIsPattern(
+                this.Compilation, pattern.Syntax, expression, pattern, whenTrueLabel: whenTrueLabel, whenFalseLabel: whenFalseLabel, diagnostics);
+            if (!hasErrors && !decisionDag.ReachableLabels.Contains(whenTrueLabel))
+            {
+                diagnostics.Add(ErrorCode.ERR_IsPatternImpossible, node.Location, expression.Type);
+                hasErrors = true;
             }
 
-            BoundPattern pattern = BindPattern(node.Pattern, expressionType, hasErrors, diagnostics);
-            if (!hasErrors && pattern is BoundDeclarationPattern p && !p.IsVar && expression.ConstantValue == ConstantValue.Null)
+            if (expression.ConstantValue != null)
             {
-                diagnostics.Add(ErrorCode.WRN_IsAlwaysFalse, node.Location, p.DeclaredType.Type);
+                decisionDag = decisionDag.SimplifyDecisionDagForConstantInput(expression, this.Conversions, diagnostics);
+                if (!hasErrors)
+                {
+                    if (!decisionDag.ReachableLabels.Contains(whenTrueLabel))
+                    {
+                        diagnostics.Add(ErrorCode.WRN_GivenExpressionNeverMatchesPattern, node.Location);
+                    }
+                    else if (!decisionDag.ReachableLabels.Contains(whenFalseLabel) && pattern.Kind == BoundKind.ConstantPattern)
```

Comment:
>&& pattern.Kind == BoundKind.ConstantPattern [](start = 83, length = 44)

When is the pattern not a `ConstantPattern`?

Reply:
``` c#
        const string s = null;
        System.Console.WriteLine(s is string { Length: 3 });
```